### PR TITLE
HIVE-25149: Support parallel load for Fast HT implementations

### PR DESCRIPTION
**Pull request by pgaref**

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastLongHashSetContainer.java [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+  }
+
+  @Override
+  public VectorMapJoinNonMatchedIterator createNonMatchedIterator(MatchTracker matchTracker) {
+    throw new RuntimeException("Not implemented");
+  }
+
+  @Override
+  public int spillPartitionId() {
+    throw new RuntimeException("Not implemented");
```
Comment (belugabehr) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastLongHashSetContainer.java:138]: UnsupportedOperationException

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastMultiKeyHashMapContainer.java [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+  }
+
+  @Override
+  public VectorMapJoinNonMatchedIterator createNonMatchedIterator(MatchTracker matchTracker) {
+    return new NonMatchedBytesHashMapParallelIterator(matchTracker, vectorMapJoinFastMultiKeyHashMaps, numThreads);
+  }
+
+  @Override
+  public int spillPartitionId() {
+    throw new RuntimeException("Not implemented");
```
Comment (belugabehr) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastMultiKeyHashMapContainer.java:171]: UnsupportedOperationException

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastMultiKeyHashMultiSetContainer.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    int count = 0;
+    for (int i=0; i < numThreads; ++i) {
+      count += vectorMapJoinFastMultiKeyHashMultiSets[i].logicalHashBucketCount;
+    }
+    return MatchTracker.create(count);
+  }
+
+  @Override
+  public VectorMapJoinNonMatchedIterator createNonMatchedIterator(MatchTracker matchTracker) {
+    throw new RuntimeException("Not implemented");
```
Comment (belugabehr) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastMultiKeyHashMultiSetContainer.java:105]: UnsupportedOperationException

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastMultiKeyHashMultiSetContainer.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+  }
+
+  @Override
+  public int spillPartitionId() {
+    throw new RuntimeException("Not implemented");
```
Comment (belugabehr) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastMultiKeyHashMultiSetContainer.java:110]: UnsupportedOperationException

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastMultiKeyHashSetContainer.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    int count = 0;
+    for (int i=0; i < numThreads; ++i) {
+      count += vectorMapJoinFastMultiKeyHashSets[i].logicalHashBucketCount;
+    }
+    return MatchTracker.create(count);
+  }
+
+  @Override
+  public VectorMapJoinNonMatchedIterator createNonMatchedIterator(MatchTracker matchTracker) {
+    throw new RuntimeException("Not implemented");
```
Comment (belugabehr) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastMultiKeyHashSetContainer.java:104]: UnsupportedOperationException

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastMultiKeyHashSetContainer.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+  }
+
+  @Override
+  public int spillPartitionId() {
+    throw new RuntimeException("Not implemented");
```
Comment (belugabehr) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastMultiKeyHashSetContainer.java:109]: UnsupportedOperationException

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastStringCommon.java [hunk 7]
```diff
@@ -59,10 +60,27 @@ public boolean adaptPutRow(VectorMapJoinFastBytesHashTable hashTable,
...
+    int keyLength = currentKey.getLength();
+    keyBinarySortableDeserializeRead.set(keyBytes, 0, keyLength);
+    try {
+      if (!keyBinarySortableDeserializeRead.readNextField()) {
+        return 0;
+      }
+    } catch (Exception e) {
+      throw new HiveException(
+          "\nDeserializeRead details: " + keyBinarySortableDeserializeRead.getDetailedReadPositionString()
+              + "\nException: " + e.toString());
```
Comment (belugabehr) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastStringCommon.java:78]: Ditch the new-lines, wrap the Exception

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastStringHashMapContainer.java [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+  }
+
+  @Override
+  public VectorMapJoinNonMatchedIterator createNonMatchedIterator(MatchTracker matchTracker) {
+    return new NonMatchedBytesHashMapIterator(matchTracker, vectorMapJoinFastStringHashMaps, numThreads);
+  }
+
+  @Override
+  public int spillPartitionId() {
+    throw new RuntimeException("Not implemented");
```
Comment (belugabehr) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastStringHashMapContainer.java:189]: UnsupportedOperationException

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastStringHashMultiSetContainer.java [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    int count = 0;
+    for (int i=0; i < numThreads; ++i) {
+      count += vectorMapJoinFastStringHashMultiSets[i].logicalHashBucketCount;
+    }
+    return MatchTracker.create(count);
+  }
+
+  @Override
+  public VectorMapJoinNonMatchedIterator createNonMatchedIterator(MatchTracker matchTracker) {
+    throw new RuntimeException("Not implemented");
```
Comment (belugabehr) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastStringHashMultiSetContainer.java:123]: UnsupportedOperationException

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastStringHashMultiSetContainer.java [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+  }
+
+  @Override
+  public int spillPartitionId() {
+    throw new RuntimeException("Not implemented");
```
Comment (belugabehr) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastStringHashMultiSetContainer.java:128]: UnsupportedOperationException

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastStringHashSetContainer.java [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+    byte[] keyBytes = currentKey.getBytes();
+    int keyLength = currentKey.getLength();
+    keyBinarySortableDeserializeRead.set(keyBytes, 0, keyLength);
+    try {
+      if (!keyBinarySortableDeserializeRead.readNextField()) {
+        return 0;
+      }
+    } catch (Exception e) {
+      throw new HiveException("\nDeserializeRead details: " +
+          keyBinarySortableDeserializeRead.getDetailedReadPositionString() + "\nException: " + e);
```
Comment (belugabehr) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastStringHashSetContainer.java:86]: Ditch the new-lines, wrap the exception

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastStringHashSetContainer.java [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+    int count = 0;
+    for (int i=0; i < numThreads; ++i) {
+      count += vectorMapJoinFastStringHashSets[i].logicalHashBucketCount;
+    }
+    return MatchTracker.create(count);
+  }
+
+  @Override
+  public VectorMapJoinNonMatchedIterator createNonMatchedIterator(MatchTracker matchTracker) {
+    throw new RuntimeException("Not implemented");
```
Comment (belugabehr) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastStringHashSetContainer.java:123]: UnsupportedOperationException

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastStringHashSetContainer.java [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+  }
+
+  @Override
+  public int spillPartitionId() {
+    throw new RuntimeException("Not implemented");
```
Comment (belugabehr) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastStringHashSetContainer.java:128]: UnsupportedOperationException

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastStringHashSetContainer.java [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+      VectorMapJoinHashSetResult hashSetResult) throws IOException {
+    long hashCode = HashCodeUtil.murmurHash(keyBytes, keyStart, keyLength);
+    return vectorMapJoinFastStringHashSets[(int) ((numThreads - 1) & hashCode)].contains(keyBytes, keyStart, keyLength, hashSetResult);
+  }
+
+  @Override
+  public VectorMapJoinHashSetResult createHashSetResult() {
+    return new VectorMapJoinFastBytesHashSetStore.HashSetResult();
+  }
+}
```
Comment (belugabehr) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastStringHashSetContainer.java:142]: Add new line

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastTableContainer.java [hunk 15]
```diff
@@ -233,20 +221,34 @@ public boolean hasSpill() {
...
+  }
+
+  @Override
+  public VectorMapJoinNonMatchedIterator createNonMatchedIterator(MatchTracker matchTracker) {
+    throw new RuntimeException("Not applicable");
+  }
+
+  @Override
+  public int spillPartitionId() {
+    throw new RuntimeException("Not applicable");
```
Comment (belugabehr) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastTableContainer.java:246]: UnsupportedOperationException

**Review by belugabehr — CHANGES_REQUESTED** [review]

**Comment by pgaref** [comment]
Thanks for the comments @belugabehr addressing them shortly!
Just keep in mind that LegacyVectorMapJoinFastHashTableLoader is a JMH class only used for benchmarking purposes -- it is merely a copy of the existing VectorMapJoinFastHashTableLoader class to make sure that there is no regression introduced.

Cheers

File: itests/hive-jmh/src/main/java/org/apache/hive/benchmark/vectorization/mapjoin/load/LegacyVectorMapJoinFastHashTableLoader.java [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+import org.apache.tez.runtime.library.api.KeyValueReader;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Map;
+
+public class LegacyVectorMapJoinFastHashTableLoader implements org.apache.hadoop.hive.ql.exec.HashTableLoader {
+  private static final Logger LOG = LoggerFactory.getLogger(LegacyVectorMapJoinFastHashTableLoader.class.getName());
```
[the remark this answers is not in the document]
Comment (pgaref) [reply itests/hive-jmh/src/main/java/org/apache/hive/benchmark/vectorization/mapjoin/load/LegacyVectorMapJoinFastHashTableLoader.java:47]: ack

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastStringHashSetContainer.java (hunk 11 above)
Comment (pgaref) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastStringHashSetContainer.java:86]: sure

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastStringHashSetContainer.java (hunk 12 above)
Comment (pgaref) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastStringHashSetContainer.java:123]: ack

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastStringHashSetContainer.java (hunk 13 above)
Comment (pgaref) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastStringHashSetContainer.java:128]: ack

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/optimized/VectorMapJoinOptimizedHashMap.java [hunk 17]
```diff
@@ -31,6 +31,7 @@
 import org.apache.hadoop.hive.ql.exec.vector.mapjoin.hashtable.VectorMapJoinNonMatchedIterator;
 import org.apache.hadoop.hive.ql.metadata.HiveException;
 import org.apache.hadoop.hive.serde2.WriteBuffers.ByteSegmentRef;
+import org.apache.hadoop.io.BytesWritable;
```
Comment (belugabehr) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/optimized/VectorMapJoinOptimizedHashMap.java:34]: This seems like an artifact of your development. Please remove.

Comment (pgaref) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/optimized/VectorMapJoinOptimizedHashMap.java:34]: yup, thanks!

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastHashTableLoader.java [hunk 18]
```diff
@@ -73,6 +104,74 @@ public void init(ExecMapperContext context, MapredContext mrContext,
...
+      throws InterruptedException, IOException, SerDeException {
+    for (int partitionId = 0; partitionId < numLoadThreads; partitionId++) {
+      int finalPartitionId = partitionId;
+      this.loadExecService.submit(() -> {
+        try {
+          LOG.info("Partition id {} with Queue size {}", finalPartitionId, loadBatchQueues[finalPartitionId].size());
+          drainAndLoadForPartition(finalPartitionId, vectorMapJoinFastTableContainer);
+        } catch (IOException | InterruptedException | SerDeException | HiveException e) {
+          LOG.error("Failed to start HT Load threads! {}", e);
+          throw new RuntimeException(e.getMessage(), e);
```
[the remark this answers is not in the document]
Comment (pgaref) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastHashTableLoader.java:148]: ack

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastHashTableLoader.java [hunk 19]
```diff
@@ -139,50 +237,72 @@ public void load(MapJoinTableContainer[] mapJoinTables,
...
-            vectorMapJoinFastTableContainer.getEstimatedMemorySize(), delta);
+              "estimatedMemoryUsage: {} Load Time : {} ", inputName, cacheKey, receivedEntries,
+            tableContainer.getEstimatedMemorySize(), delta);
         } else {
           LOG.info("Finished loading hash table for input: {} cacheKey: {} numEntries: {} Load Time : {} ",
-                  inputName, cacheKey, numEntries, delta);
+                  inputName, cacheKey, receivedEntries, delta);
         }
+      } catch (InterruptedException e) {
+        throw new HiveException(e);
```
[the remark this answers is not in the document]
Comment (pgaref) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastHashTableLoader.java:317]: ack

**Pull request closed without merging**